proposals/unions.md

```diff
@@ -0,0 +1,322 @@
+
+Even when a union switch is otherwise exhaustive, if the null state of the incoming union's `Value` property is "maybe null", a warning will be given on unhandled null.
+
+``` c#
+Pet pet = GetNullableDog(); // 'pet.Value' is "maybe null"
+var value = pet switch
+{
+    Dog dog => ...,
+    Cat cat => ...,
+    // Warning: 'null' not handled
+}
+```
+
+### Union declarations
+
+Union declarations are a succinct and opinionated way of declaring union types in C#. They declare a struct which uses a single object reference for storage, which means that:
+
+* Any value types among their case types will be boxed on entry.
+* Union values are pointer-length in size, and thus protected from "tearing".
+
+The intent is for union declarations to cover the vast majority of use cases quite nicely. The two main reasons for hand coding specific union types rather than use union declarations are expected to be:
+
+* Adapting existing types to the union patterns to gain union behaviors.
+* Implementing a different storage strategy for e.g. efficiency or interop reasons.
+
+#### Syntax
+
+A union declaration has a name and a list of case types.
+
+``` antlr
+union_declaration
+    : attributes? struct_modifier* 'partial'? 'union' identifier type_parameter_list?
+      '(' type (',' type)* ')' type_parameter_constraints_clause* 
+      (`{` struct_member_declaration* `}` | ';')
+    ;
+```
+
+In addition to the restrictions on struct members ([§16.3](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/structs.md#163-struct-members)), the following applies to union members:
+
+* Instance fields, auto-properties or field-like events are not permitted.
+* Explicitly declared public constructors with a single parameter are not permitted.
+* Explicitly declared constructors must use a `this(...)` initializer to (directly or indirectly) delegate to one of the generated constructors.
+
+Case types can be any type that converts to `object`, e.g., interfaces, type parameters, nullable types and other unions. It is fine for cases to overlap, and for unions to nest or be null.
+
+#### Lowering
+
+A union declaration is lowered to a struct declaration with
+
+* the same attributes, modifiers, name, type parameters and constraints,
+* implicit implementations of `IUnion` and `IUnion<T>`,
+* a `public object? Value { get; }` auto-property,
+* a public constructor for each case type,
+* any members in the union declaration's body.
+
+If any user-declared members conflict with generated members, only the user-declared members are kept.
+
+Example:
+
+``` c#
+public union Pet(Cat, Dog){ ... }
+```
+
+Is lowered to:
+
+``` c#
+public record struct Pet : IUnion, IUnion<Pet>
```

Comment:
maybe call out somewhere this means that unions have well-defined contracts around equality/hashing/==/!= from `record`.

Reply:
I don't know if we agreed that unions would lower to records though. I'm not against it.

Reply:
Good point, I added that. I also went ahead and changed the `Value` property to have an `init` accessor, so that `with` expressions will work. Not sure if that's worthwhile, but it seems harmless.